### ISSUE TITLE
Add ShadowDOM support from WHATWG-DOM

### ISSUE THREAD

**Issue by RByers** [w3c/dom#13]
[Pointer Events](https://w3c.github.io/pointerevents/) needs to refer to the ShadowDOM support in the DOM spec (eg. the [composed property](https://dom.spec.whatwg.org/#dom-event-composed)) and so currently references WHATWG-DOM.

If W3C DOM is updated to match implementations (and kept up-to-date) on things like this, then we could reference DOM4 instead of WHATWG-DOM.  PEWG is [highly test-driven](https://github.com/w3c/pointerevents#editing-process), so when WHATWG-DOM more accurately represents the behavior we're covering in our tests, we'll have to reference it instead in order to keep our tests green against nightly chromium and firefox builds (the tests are run automatically in our infrastructure and used to prevent regressions on ever blink/gecko check-in).

**Comment by chaals** [comment]
Hi Rick,

It appears from the Pointer Events spec that what you use from Shadow DOM is the reference to the `composed` attribute. Is there anything else?

The tests for this are in the Shadow DOM repo, right? So if I understand correctly (which is far from certain) this is basically implemented in shipping webkit and blink, right?

How much of that should be included in the DOM spec directly, how much in the Shadow DOM spec? 

ping @hayatoito @yongsheng

**Comment by yongsheng** [comment]
Currently I see Shadow DOM is an independent draft. Need to check whether it should be released standalone or included in DOM spec. I'm neutral about this. @chaals do you know anything about shadow dom working team?

**Comment by RByers** [comment]
> It appears from the Pointer Events spec that what you use from Shadow DOM is the reference to the composed attribute. Is there anything else?

Yeah it looks like that's all we reference explicitly.  

> The tests for this are in the Shadow DOM repo, right? So if I understand correctly (which is far from certain) this is basically implemented in shipping webkit and blink, right?

Yes, and behind a flag in Gecko.  It looks like we have a test hole in WPT here - what I was remembering as WPT tests are actually only blink LayoutTests 😢, sorry.  I'll look into filling that separately.  In the meantime this should be the same as UI Events.  As long as we can point to some `composed` definition somewhere (doesn't matter to me if it's in DOM or a separate spec) then that should be tolerable.

**Comment by yongsheng** [comment]
It seems there are no other implementations for shadow dom.
https://developer.mozilla.org/en-US/docs/Web/Web_Components/Shadow_DOM
Need a double check.

**Comment by TakayoshiKochi** [comment]
@yongsheng It looks like MDN docs are quite out of date and needs updates. Please don't refer MDN documents at this moment.

There is an independent [Shadow DOM spec](https://w3c.github.io/webcomponents/spec/shadow/), but it is  being upstreamed to WHATWG DOM spec and WHATWG HTML spec. So WHATWG one is more accurate and up to date, and W3C DOM spec lags behind.

**Comment by yongsheng** [comment]
yes, it's outdated. I mean the implementation status of shadow dom. seems only chromium supports it. anyone who knows the impl status of firefox?

**Comment by TakayoshiKochi** [comment]
I guess Mozilla people are working on implementation, but am not sure.
Chrome (Blink) and Safari (WebKit) implement Shadow DOM.

I'm not sure what is your concern here about implementation status, though.
Could you elaborate?

**Comment by yongsheng** [comment]
Not finalized yet. will check this with other specs, like HTML.

**Comment by chaals** [comment]
Maybe I can elaborate. The idea with the DOM spec, like the HTML spec, is to put stuff in that will get through CR. We haven't tried to add things that we expect in the future, and last time I talked directly to @yongsheng and others working on the spec, we were not sure that Shadow DOM would have more than Chromium implementation within the timeline for this version, and there was a W3C Shadow DOM spec so we deferred the question.

If reality suggests that we should be taking it into DOM 4.1, then we'll work out how to do that.
